base/dates/parse.jl

```diff
@@ -1,97 +1,132 @@
 ### Parsing utilities
 
-@generated function tryparse_internal{T<:TimeType, S, F}(::Type{T}, str::AbstractString, df::DateFormat{S, F}, raise::Bool=false)
-    token_types = Type[dp <: DatePart ? SLOT_RULE[first(dp.parameters)] : Void for dp in F.parameters]
-    N = length(F.parameters)
-
-    types = slot_order(T)
-    num_types = length(types)
-    order = Vector{Int}(num_types)
-    for i = 1:num_types
-        order[i] = findfirst(token_types, types[i])
+function directives{S,F}(::Type{DateFormat{S,F}})
+    tokens = F.parameters
+    di = 1
+    directive_index = zeros(Int, length(tokens))
+    directive_letters = sizehint!(Char[], length(tokens))
+    for (i, token) in enumerate(tokens)
+        if token <: DatePart
+            directive_index[i] = di
+
+            letter = first(token.parameters)
+            push!(directive_letters, letter)
+
+            di += 1
+        end
     end
+    return tokens, directive_index, directive_letters
+end
+
+genvar(t::DataType) = Symbol(lowercase(string(t.name.name)))
```

Comment:
I'm sure this is going to generate lots of discussion. Basically I need to map the specifier types to a variable name. Using separate variable names allows the generated code to avoid type instability problems (occurs once extended). Additionally, I can't get away with using `gensym` as some of the code requires me to match up the variable names. See `tryparse_internal`